### PR TITLE
feat(editor): editor object in place of store exports

### PR DESCRIPTION
**Pull request by hejtful**
**Change in API design**: Instead of providing only the rendered `editor` element, provide an `editor` object containing the element, language data, Redux store data, etc.

**Localized change**: To reduce the amount of Editor exports, provide the Redux store data (`history` and `selectRootDocument`) as properties of the `editor` object. These properties only expose the data from the store needed for using the Editor package.

**Side effect changes**: Split the `packages/editor/src/core/editor.tsx` file, that held two components, into three files that hold one component each. Extracted types into a separate file. Rename `languageData` property to `i18n`.

Matching edusharing PR: https://github.com/serlo/serlo-editor-for-edusharing/pull/361

### CODE REVIEW

File: packages/editor/src/core/types.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  initialState: {
+    plugin: string
+    state?: unknown
+  }
+  onChange?: OnEditorChange
+}
+
+export type EditorChildren = ReactNode | ((editor: EditorData) => ReactNode)
+
+export interface EditorData {
```
Comment (hejtful) [inline packages/editor/src/core/types.ts]: A better name, maybe? 😬 Any suggestions?

File: packages/editor/src/core/editor-children.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        pendingChanges,
+        dispatchUndo,
+        dispatchRedo,
+        dispatchPersistHistory,
+      },
+      selectRootDocument,
+    })
+  }
+
+  return (
```
Comment (shn-srl) [inline packages/editor/src/core/editor-children.tsx]: nitpick: Returning this above the if statement would be better (return early)

File: packages/editor/src/core/types.ts (hunk 1 above)
Comment (shn-srl) [reply packages/editor/src/core/types.ts]: EditorConfigData ? :see_no_evil:

**Review by shn-srl — APPROVED** [review]

File: packages/editor/src/core/editor-children.tsx (hunk 2 above)
Comment (LarsTheGlidingSquirrel) [reply packages/editor/src/core/editor-children.tsx]: Could we drop this case altogether? I think all editor integrations will want to pass a function (render prop) as `children` to get access to `selectRootDocument`, which is needed to save the document.

File: packages/editor/src/core/types.ts (hunk 1 above)
Comment (LarsTheGlidingSquirrel) [reply packages/editor/src/core/types.ts]: SerloEditorChildrenParameters ? (or abbreviated)

File: packages/editor/src/core/editor-children.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    void dispatch(persistHistory(documents))
+  }, [dispatch])
+
+  const selectRootDocument = useCallback(() => {
+    return selectStaticDocument(store.getState(), ROOT)
+  }, [])
+
+  const editor = <SubDocument id={ROOT} />
+
+  if (typeof children === 'function' && loggedInData) {
```
Comment (LarsTheGlidingSquirrel) [inline packages/editor/src/core/editor-children.tsx]: As far as I understand `loggedInData` can never be falsy here. Maybe clarify that in the code?

**Review by LarsTheGlidingSquirrel — APPROVED** [review]
Thank you! Added some comments but no blockers from my point of view.

File: packages/editor/src/core/editor-children.tsx (hunk 2 above)
Comment (hejtful) [reply packages/editor/src/core/editor-children.tsx]: @shn-srl good point!

@LarsTheGlidingSquirrel The frontend does use this case though.

File: packages/editor/src/core/editor-children.tsx (hunk 3 above)
Comment (hejtful) [reply packages/editor/src/core/editor-children.tsx]: I'm not sure what you're saying, sorry. Its type at this point is `LoggedInData | null`, therefore this check. Maybe you could provide like a code block suggestion of the changes you're proposing?

Comment (LarsTheGlidingSquirrel) [reply packages/editor/src/core/editor-children.tsx]: Something like 
```
if(!loggedInData) { 
  // Handle error, maybe return null, not sure what would be best here 
}
if (typeof children === 'function') {
  // ...
}
```
Right now, it reads like also `loggedInData` determines if we call `children` or render it directly. 
Hope that makes it more clear.

File: packages/editor/src/core/editor-children.tsx (hunk 2 above)
Comment (LarsTheGlidingSquirrel) [reply packages/editor/src/core/editor-children.tsx]: > @LarsTheGlidingSquirrel The frontend does use this case though.

Oh, you are right. Nevermind.

File: packages/editor/src/core/editor-children.tsx (hunk 3 above)
Comment (hejtful) [reply packages/editor/src/core/editor-children.tsx]: Thanks for explaining, makes sense :)

**Pull request merged**